In [1]:
from jw_package import *

import configuration completed !
Data configuration has been started !
Data configuration completed !
function configuration completed ! 
Good to go !


### $$ A*w = b$$
    * A = M X N matrix that row = tuples, column = features
    * w = N vector that weight for each feature
    * b = M vector that label for desired 'units'

1. w는 item_nbr에 대한 feature들의 weight다.
2. 그럼 w는 item_nbr별로 설정해야 할까?
3. store마다 환경요인이 다를 수 있으므로 store_nbr별 item_nbr에 대한 w를 각기 설정해야 정확한 예측이 될 것 같다
4. station이 같은 store라면 완전히 같진 않더라도 아마 비슷한 경향을 보일듯
5. store_nbr별 item_nbr에 대한 w를 구하기로 했다면 총 몇 개를 구해야 할까?
6. store당 팔리는 item 종류가 평균적으로 5개라 치면 45 * 5 = 225, w가 대략 225개 만들어진다.
7. 당연히 일일이 구하진 않을거고 상관계수가 높은 애들만 추려서 가중치 부여하는 식으로 w를 구함
8. 근데 store_nbr별로 item_nbr에 대한 w를 각기 구하는게 잘하는 걸까?
9. item_nbr별로만 하면 안되나? 내 생각엔 안 될 것 같다.
10. 예를 들어 어떤 item이 더운 곳에선 잘 팔린다면 같은 시기에 추운 곳에선 당연히 잘 안 팔릴 것이지 않은가
11. 그럼 item_nbr별로 같은 w를 부여하면 더운 곳, 추운 곳을 가려서 제대로 예측을 못 하겠지?
12. 그러니까 store_nbr별 item_nbr의 w가 맞는 것 같다
## 결론 = 위 방정식에 쓰일 w는 store_nbr별 item_nbr에 대해 각기 구해야 하지 않을까?

# 그럼 방정식에서 기본값은 뭐로?
# 기본값 크기는 어떻게? (store_nbr wise)

# 전체 store_nbr별 item_nbr의 총 갯수는?

In [2]:
total = 0
for store in np.arange(1,46):
    item_nbrs = identify_item_nbr(store)
    total += len(item_nbrs)
total

203

# 총 203개의 w벡터..
## 그럼 store별 item을 예측할 때 어떤 feature에 가중치를 얼마나 줄 지 어떻게 알까..
## 일단 1번 store를 기준으로 출력해보자

In [3]:
store_1_2012 = get_correlation(1,2012)
store_1_2013 = get_correlation(1,2013)
store_1_2014 = get_correlation(1,2014)

C:\Users\WJ\Anaconda3\lib\site-packages\scipy\stats\stats.py:2998: RuntimeWarning: Mean of empty slice.
  mx = x.mean()
C:\Users\WJ\Anaconda3\lib\site-packages\numpy\core\_methods.py:80: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)
C:\Users\WJ\Anaconda3\lib\site-packages\scipy\stats\stats.py:2999: RuntimeWarning: Mean of empty slice.
  my = y.mean()
C:\Users\WJ\Anaconda3\lib\site-packages\scipy\stats\stats.py:3003: RuntimeWarning: invalid value encountered in double_scalars
  r = r_num / r_den
C:\Users\WJ\Anaconda3\lib\site-packages\scipy\stats\stats.py:5240: RuntimeWarning: invalid value encountered in less
  x = np.where(x < 1.0, x, 1.0)  # if x > 1 then return 1.0


In [6]:
show_me_pearson(store_1_2012)

(1, 2012, -1, -2)
	 ((9, 'avgspeed', 364), (0.029413892409414884, 0.5759077562060245))
	 ((9, 'cool', 362), (-0.06507347742720462, 0.21678034409391772))
	 ((9, 'depart', 0), (nan, nan))
	 ((9, 'dewpoint', 363), (-0.16263173725428326, 0.0018798601109756173))
	 ((9, 'heat', 362), (0.1962103722989204, 0.00017208151618507012))
	 ((9, 'preciptotal', 303), (-0.07761708837800624, 0.17781222909816158))
	 ((9, 'resultdir', 363), (0.05088556611367485, 0.33365257830459105))
	 ((9, 'resultspeed', 363), (0.048113484264033224, 0.36068909407025174))
	 ((9, 'sealevel', 361), (0.10930797729412822, 0.03790572490368731))
	 ((9, 'snowfall', 0), (nan, nan))
	 ((9, 'stnpressure', 363), (0.10580652394994502, 0.04394758191932755))
	 ((9, 'sunrise', 0), (nan, nan))
	 ((9, 'sunset', 0), (nan, nan))
	 ((9, 'tavg', 362), (-0.1777482034736896, 0.0006805793222412667))
	 ((9, 'tmin', 364), (-0.1657085413387121, 0.00151077966541003))
	 ((9, 'wetbulb', 360), (-0.16325684039123894, 0.0018855999939363544))
	 ((28, 'avgs

# 너무 어지럽다
# 참고로 중간에 nan값은 column이 통째로 M인 경우이다
# 이제 상관관계가 높은 (item_nbr, feature)들만 filtering해서 보자

In [7]:
trimmed_store_1_2012 = trim_correlation(store_1_2012)
show_me_pearson(trimmed_store_1_2012)

(1, 2012, -1, -2)
	 ((99, 'cool', 362), (0.23360735334567914, 7.068685749589456e-06))
	 ((99, 'wetbulb', 360), (0.22639519766547406, 1.4440842669249175e-05))
	 ((99, 'dewpoint', 363), (0.22037814773777978, 2.2704690497583713e-05))
	 ((99, 'tmin', 364), (0.21596943064827212, 3.247128857492015e-05))
	 ((99, 'tavg', 362), (0.2140769226856364, 4.013315906375752e-05))
	 ((99, 'heat', 362), (-0.18171028709217207, 0.0005122189784035525))
	 ((40, 'heat', 362), (0.3499211026685748, 7.2561712206726976e-12))
	 ((40, 'tavg', 362), (-0.3348977719122911, 6.167367009299315e-11))
	 ((40, 'tmin', 364), (-0.305428266612124, 2.6825967727965087e-09))
	 ((40, 'wetbulb', 360), (-0.2982182391821733, 7.906522553245026e-09))
	 ((40, 'dewpoint', 363), (-0.25763556532294596, 6.493338474359796e-07))
	 ((40, 'cool', 362), (-0.17878773214082763, 0.0006320391888293028))
	 ((9, 'heat', 362), (0.1962103722989204, 0.00017208151618507012))
	 ((9, 'tavg', 362), (-0.1777482034736896, 0.0006805793222412667))
	 ((9, 'tmin',

# 99번 item은 cool, wetbulb, dewpoint, tmin, tavg.. 등 위 표에서 확인할 수 있는 것과 같이 유의미한 상관관계가 있음을 확인할 수 있다

# 단, 모든 store, year에서가 아니라 2012년의 1번 store에서 그렇당